lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1672 @@
+		let proportional_millionths: u32 = Readable::read(reader)?;
+		Ok(RoutingFees {
+			base_msat,
+			proportional_millionths,
+		})
+	}
+}
+
+impl Writeable for RoutingFees {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		self.base_msat.write(writer)?;
+		self.proportional_millionths.write(writer)?;
+		Ok(())
+	}
+}
+
+#[derive(PartialEq, Debug)]
+/// Information received in the latest node_announcement from this node.
+pub struct NodeAnnouncementInfo {
+	/// Protocol features the node announced support for
+      pub features: NodeFeatures,
+	/// When the last known update to the node state was issued
+      pub last_update: u32,
+	/// Color assigned to the node
+	pub rgb: [u8; 3],
+	/// Moniker assigned to the node
+	pub alias: [u8; 32],
+	/// Internet-level addresses via which one can connect to the node
+	pub addresses: Vec<NetAddress>,
+	/// An initial announcement of the node
+	// this is cached here so we can send out it later if required by initial routing sync
+	// keep an eye on this to see if the extra memory is a problem
+	pub announcement_message: Option<msgs::NodeAnnouncement>
+}
+
+impl Writeable for NodeAnnouncementInfo {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		self.features.write(writer)?;
+		self.last_update.write(writer)?;
+		self.rgb.write(writer)?;
+		self.alias.write(writer)?;
+		(self.addresses.len() as u64).write(writer)?;
+		for ref addr in &self.addresses {
+			addr.write(writer)?;
+		}
+		self.announcement_message.write(writer)?;
+		Ok(())
+	}
+}
+
+impl Readable for NodeAnnouncementInfo {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NodeAnnouncementInfo, DecodeError> {
+		let features = Readable::read(reader)?;
+		let last_update = Readable::read(reader)?;
+		let rgb = Readable::read(reader)?;
+		let alias = Readable::read(reader)?;
+		let addresses_count: u64 = Readable::read(reader)?;
+		let mut addresses = Vec::with_capacity(cmp::min(addresses_count, MAX_ALLOC_SIZE / 40) as usize);
+		for _ in 0..addresses_count {
+			match Readable::read(reader) {
+				Ok(Ok(addr)) => { addresses.push(addr); },
+				Ok(Err(_)) => return Err(DecodeError::InvalidValue),
+				Err(DecodeError::ShortRead) => return Err(DecodeError::BadLengthDescriptor),
+				_ => unreachable!(),
+			}
+		}
+		let announcement_message = Readable::read(reader)?;
+		Ok(NodeAnnouncementInfo {
+			features,
+			last_update,
+			rgb,
+			alias,
+			addresses,
+			announcement_message
+		})
+	}
+}
+
+#[derive(PartialEq)]
+/// Details regarding a node in the network
+pub struct NodeInfo {
+	/// All valid channels a node has announced
+	pub channels: Vec<u64>,
+	/// Lowest fees enabling routing via any of the known channels to a node
+	pub lowest_inbound_channel_fees: Option<RoutingFees>,
+	/// More information about a node from node_announcement
+	/// Optional because we may have a NodeInfo entry before having received the announcement
```

Comment:
*do* have :p. Its actually illegal to receive node_announcement before having received channels for the node (which will create this).